test/new-e2e/utils/e2e/e2e.go

```diff
@@ -3,8 +3,218 @@
+//	}
+//
+// In this example, the components available are defined by the struct [e2e.VMEnv] which contains a virtual machine.
+// The generic type of [e2e.Suite] must match the type of the stack definition.
+// In our example, [e2e.EC2VMStackDef] returns an instance of [*e2e.StackDefinition][[e2e.VMEnv]].
+//
+// The following default stack definitions are provided:
+//   - [e2e.EC2VMStackDef] creates an environment with a virtual machine. See [e2e.EC2VMStackDef] for more information about the supported options.
+//   - [e2e.AgentStackDef] creates an environment with an Agent installed on a virtual machine. See [e2e.AgentStackDef] for more information about the supported options.
+//
+// # Defining your stack definition
+//
+// In some special cases, you have to define a custom environment.
+// Here is an example on how to define an environment with Docker installed on a virtual machine.
+//
+//	type dockerSuite struct {
+//		e2e.Suite[e2e.VMEnv]
+//	}
+//
+//	func TestDockerSuite(t *testing.T) {
+//		e2e.Run(t, &dockerSuite{}, e2e.EnvFactoryStackDef(dockerEnvFactory))
+//	}
+//
+//	func dockerEnvFactory(ctx *pulumi.Context) (*e2e.VMEnv, error) {
+//		vm, err := ec2vm.NewUnixEc2VM(ctx)
+//		if err != nil {
+//			return nil, err
+//		}
+//
+//		_, err = docker.NewAgentDockerInstaller(vm.UnixVM, docker.WithAgent(docker.WithAgentImageTag("7.42.0")))
+//
+//		if err != nil {
+//			return nil, err
+//		}
+//
+//		return &e2e.VMEnv{
+//			VM: client.NewVM(vm),
+//		}, nil
+//	}
+//
+//	func (docker *dockerSuite) TestDocker() {
+//		docker.Env().VM.Execute("docker container ls")
+//	}
+//
+// [e2e.EnvFactoryStackDef] is used to define a custom environment.
+// Here is a non exhaustive list of components that can be used to create a custom environment:
+//   - [EC2 VM]: Provide methods to create a virtual machine on EC2.
+//   - [Agent]: Provide methods to install the Agent on a virtual machine
+//   - [File Manager]: Provide methods to manipulate files and folders
+//
+// # Organizing your tests
+//
+// # Having a single environment
+//
+// In the simple case, there is a single environment and each test checks one specific thing.
+//
+//	type singleEnvSuite struct {
+//		e2e.Suite[e2e.AgentEnv]
+//	}
+//
+//	func TestSingleEnvSuite(t *testing.T) {
+//		e2e.Run(t, &singleEnvSuite{}, e2e.AgentStackDef(nil))
+//	}
+//
+//	func (suite *singleEnvSuite) Test1() {
+//		// Check feature 1
+//	}
+//
+//	func (suite *singleEnvSuite) Test2() {
+//		// Check feature 2
+//	}
+//
+//	func (suite *singleEnvSuite) Test3() {
+//		// Check feature 3
+//	}
+//
+// # Having different environments
+//
+// In this scenario, the environment is different for each test (or for most of them).
+// [e2e.Suite.UpdateEnv] is used to update the environment.
+// Keep in mind that using [e2e.Suite.UpdateEnv] to update virtual machine settings can destroy
+// the current virtal machine and create a new one when updating the operating system for example.
```

Comment:
```suggestion
// the current virtual machine and create a new one when updating the operating system for example.
```

We should also mention that it is not recommended

Reply:
I don't think so as there is a valid use case.